### ISSUE TITLE
Unintended reformatting of strings containing .or. or .and. passed to the macro compiler in XS 2.8b

### ISSUE THREAD

**Issue by ArneOrtlinghaus** [X-Sharp/XSharpPublic#694]
The string "if u < 57600.and. u > 60000 " including the double or single quotes passed to the macro compiler is interpreted as macro code and is returned as 
if u < 57600 .and. u > 60000 
and not as 
if u < 57600.and. u > 60000 
The same happens with 
if u < 57600.or. u > 60000
The macro compiler does not see that a string is passed, that should be returned as a string. 
Same with a more complex expression like 
"if u < 57600.and. u > 60000 " + "x "

**To Reproduce**
   local c as string
   local ocb as _CodeBlock  
c := '"if u < 57600.and. u > 60000 "'
Console.WriteLine("init: " + c)

         c := "{||" + c + "}"
         	ocb   := &( c )
             c := ocb:Eval()

**Expected behavior**
The macro compiler should understand what is a string constant and what is macro code and make reformattings only for macro code.

**Comment by RobertvanderHulst** [comment]
Arne,
I am not sure if there are missing string delimiters inside your strings in the above example.
Can you attach an example please ?

**Comment by ArneOrtlinghaus** [comment]
Robert, 

yes, here is the example. It is at the beginning of the file attached: 
function Testmacrostring()
   local c as string
   local ocb as _CodeBlock  
c := '"if u < 57600.and. u > 60000 "'
Console.WriteLine("init: " + c)
         c := "{||" + c + "}"
         	ocb   := &( c )
             c := ocb:Eval()
Console.WriteLine("macro: " + c)
return nil

Arne

[Program.zip](https://github.com/X-Sharp/XSharpPublic/files/6675390/Program.zip)

**Comment by RobertvanderHulst** [comment]
Arne,
I am glad you attached the example. On these github pages the nested quotes are not visible if you do not embed them in code markers with the `<>` buttons from the toolbar . `'"This works"'`
Just out of curiosity: what is the problem that you are seeing, apart from the fact that you would not expect the contents of strings inside macros to be touched. These strings look like macros as well, so sooner or later these will have to be macro compiler and then adjusted. Right ?

**Comment by ArneOrtlinghaus** [comment]
Robert, 
we use the macro compiler for many customer specific extensions and have built on top of it with many extensions. There should be more than 1 million of macro code lines over all of our installations.  

Although the risk is low there could be a case where a macro uses string constants with  sequences, the macro compiler formats because he thinks that this is macro code. I don't know all rules when the macro compiler changes the code passed to.  

I realized the change of the behavior only by a unit test that tests some "macro code lines treated as strings". So the case I have found is not a problem. But probably this is not always the case. May be there are some texts that do not represent macro code although they have character sequences that could be part of a macro expression. So in my opinion this has not been implemented correctly. 

Arne

**Comment by cpyrgas** [comment]
This problem is fixed, but while the following sample was working well before, now it throws an macrocompiler exception "Unexpected 'and'". And since such macros are very common especially in filter expressions, I am afraid this is likely to break existing code.

Maybe it's better to revert this change for now, and after the new build is out, implement a parser that adjusts strings in the macro expression only if they are not inside sub strings and send an update to Arne? 


```
FUNCTION Start() AS VOID
LOCAL c AS STRING
c := '1<2.and.2>1'
? &(c) // XSharp.MacroCompiler.CompilationError
```

**Comment by RobertvanderHulst** [comment]
Fixes in 72a763522ba68277b9111f20e3888e293e273b13

**Comment by cpyrgas** [comment]
Nice, confirmed fixed!
